model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -0,0 +1,504 @@
+		// but log if there was an error reading an exsting file
+		if !errors.Is(err, os.ErrNotExist) {
+			log.Error(err, "WARNING: could not initialize model config from file, will contine with empty config", "filename", configFilename)
+		}
+	} else {
+		var modelRepositoryConfig OvmsMultiModelRepositoryConfig
+		if err := json.Unmarshal(configBytes, &modelRepositoryConfig); err != nil {
+			log.Error(err, "WARNING: could not parse model config JSON, will contine with empty config", "filename", configFilename)
+		} else {
+			multiModelConfig = make(map[string]OvmsMultiModelConfigListEntry, len(modelRepositoryConfig.ModelConfigList))
+			for _, mc := range modelRepositoryConfig.ModelConfigList {
+				multiModelConfig[mc.Config.Name] = mc
+			}
+		}
+	}
+
+	ovmsMM := &OvmsModelManager{
+		address: address,
+		// will need to be updated before being queried
+		cachedModelConfigResponse: OvmsConfigResponse{},
+		client: &http.Client{
+			Transport: &http.Transport{
+				MaxIdleConns:        100,
+				MaxConnsPerHost:     100,
+				MaxIdleConnsPerHost: 100,
+			},
+			Timeout: 30 * time.Second,
+		},
+		log:                 log,
+		loadedModelsMap:     multiModelConfig,
+		modelConfigFilename: configFilename,
+		reqs:                make(chan *request, 25), // TODO make configurable
+	}
+
+	// write the config out on boot because OVMS needs it to exist
+	if _, err := os.Stat(configFilename); os.IsNotExist(err) {
+		if err = ovmsMM.writeConfig(); err != nil {
+			log.Error(err, "Unable to write out empty config file")
+		}
+	}
+
+	// start the actor process
+	go ovmsMM.run()
+
+	return ovmsMM
+}
+
+// TODO?
+// func (mm *OvmsModelManager) Close() {
+// 	close(mm.reqs)
+// }
+
+// "Client" API
+func (mm *OvmsModelManager) LoadModel(ctx context.Context, modelPath string, modelId string) error {
+
+	// BasePath must be a directory
+	var basePath string
+	if fileInfo, err := os.Stat(modelPath); err == nil {
+		if fileInfo.IsDir() {
+			basePath = modelPath
+		} else {
+			basePath = filepath.Dir(modelPath)
+		}
+	} else {
+		return fmt.Errorf("Could not stat file at the model_path: %w", err)
+
+	}
+
+	c := make(chan error, 1)
+	mm.reqs <- &request{
+		requestType: load,
+		modelId:     modelId,
+		basePath:    basePath,
+		c:           c,
+		ctx:         ctx,
+	}
+
+	select {
+	case err := <-c:
+		if err != nil {
+			return fmt.Errorf("LoadModel failed: %w", err)
+		}
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("LoadModel request was cancelled")
+
+	}
+}
+
+func (mm *OvmsModelManager) UnloadModel(ctx context.Context, modelId string) error {
+	c := make(chan error, 1)
+	mm.reqs <- &request{
+		requestType: unload,
+		modelId:     modelId,
+		c:           c,
+		ctx:         ctx,
+	}
+
+	select {
+	case err := <-c:
+		if err != nil {
+			return fmt.Errorf("UnloadModel failed: %w", err)
+		}
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("UnloadModel request was cancelled")
+
+	}
```

Comment:
We could consider encapsulating this submit + select in another func to eliminate a bit of duplication